Sources/NIOExtras/LengthFieldBasedFrameDecoder.swift

```diff
@@ -0,0 +1,170 @@
+///     +---+----+------+----+
+///     | A | BC | DEFG | HI |
+///     +---+----+------+----+
+///
+/// Given that the specified header length is 1 byte,
+/// where the first header specifies 3 bytes while the second header specifies 4 bytes,
+/// a `LengthFieldBasedFrameDecoder` will decode them into the following packets:
+///
+///     +-----+------+
+///     | BCD | FGHI |
+///     +-----+------+
+///
+/// 'A' and 'E' will be the headers and will not be passed forward.
+///
+
+public final class LengthFieldBasedFrameDecoder: ByteToMessageDecoder {
+    
+    ///
+    /// The decoder has two distinct sections of data to read.
+    /// Each must be fully present before it is considered as read.
+    /// During the time when it is not present the decoder must wait. `DecoderReadState` details that waiting state.
+    ///
+    private enum DecoderReadState {
+        case waitingForHeader
+        case waitingForFrame(length: Int)
+    }
+
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    
+    public var cumulationBuffer: ByteBuffer?
+    private var readState: DecoderReadState = .waitingForHeader
+    
+    private let lengthFieldLength: ByteLength
+    private let lengthFieldEndianness: Endianness
+    
+    /// Create `LengthFieldBasedFrameDecoder` with a given frame length.
+    ///
+    /// - parameters:
+    ///    - lengthFieldLength: The length of the field specifying the remaining length of the frame.
+    ///    - lengthFieldEndianness: The endianness of the field specifying the remaining length of the frame.
+    ///
+    public init(lengthFieldLength: ByteLength, lengthFieldEndianness: Endianness = .big) {
+
+        // The value contained in the length field must be able to be represented by an integer type on the platform.
+        // ie. .eight == 64bit which would not fit into the Int type on a 32bit platform.
+        precondition(lengthFieldLength.length <= Int.bitWidth/8)
+            
+        self.lengthFieldLength = lengthFieldLength
+        self.lengthFieldEndianness = lengthFieldEndianness
+    }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        
+        if case .waitingForHeader = self.readState {
+            try self.readNextLengthFieldToState(buffer: &buffer)
+        }
+        
+        guard case .waitingForFrame(let frameLength) = self.readState else {
+            return .needMoreData
+        }
+        
+        guard let frameBuffer = try self.readNextFrame(buffer: &buffer, frameLength: frameLength) else {
+            return .needMoreData
+        }
+        
+        ctx.fireChannelRead(self.wrapInboundOut(frameBuffer))
+
+        return .continue
+    }
+    
+    private func readNextLengthFieldToState(buffer: inout ByteBuffer) throws {
+        
+        guard let lengthFieldSlice = buffer.readSlice(length: self.lengthFieldLength.length) else {
```

Comment:
we're reading a `lengthFieldSlice` off here but we're never actually using it. I think you should return `Int?` here and just read off the `headerField`. Basically the same as what you have in `func frameLength` but just use `readInteger` instead of `getInteger` there.

Reply:
Agreed.